src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesCodeFixProvider.cs

```diff
@@ -0,0 +1,702 @@
+using Microsoft.CodeAnalysis.Simplification;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.RemoveUnusedParametersAndValues
+{
+    /// <summary>
+    /// Code fixer for unused expression value diagnostics reported by <see cref="AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer"/>.
+    /// We provide following code fixes:
+    ///     1. If the unused value assigned to a local/parameter has no side-effects,
+    ///        we recommend removing the assignment. We consider an expression value to have no side effects
+    ///        if one of the following is true:
+    ///         1. Value is a compile time constant.
+    ///         2. Value is a local or parameter reference.
+    ///         3. Value is a field reference with no or implicit this instance.
+    ///     2. Otherwise, if user preference is set to DiscardVariable, and project's
+    ///        language version supports discard variable, we recommend assigning the value to discard.
+    ///     3. Otherwise, we recommend assigning the value to a new unused local variable which has no reads.
+    /// </summary>
+    internal abstract class AbstractRemoveUnusedParametersAndValuesCodeFixProvider<TExpressionSyntax, TStatementSyntax, TBlockSyntax,
+                                                                                   TExpressionStatementSyntax, TLocalDeclarationStatementSyntax,
+                                                                                   TVariableDeclaratorSyntax, TForEachStatementSyntax,
+                                                                                   TSwitchCaseBlockSyntax, TSwitchCaseLabelOrClauseSyntax,
+                                                                                   TCatchStatementSyntax, TCatchBlockSyntax>
+        : SyntaxEditorBasedCodeFixProvider
+        where TExpressionSyntax : SyntaxNode
+        where TStatementSyntax : SyntaxNode
+        where TBlockSyntax : TStatementSyntax
+        where TExpressionStatementSyntax : TStatementSyntax
+        where TLocalDeclarationStatementSyntax : TStatementSyntax
+        where TVariableDeclaratorSyntax : SyntaxNode
+        where TForEachStatementSyntax: TStatementSyntax
+        where TSwitchCaseBlockSyntax: SyntaxNode
+        where TSwitchCaseLabelOrClauseSyntax: SyntaxNode
+    {
+        private static readonly SyntaxAnnotation s_memberAnnotation = new SyntaxAnnotation();
+        private static readonly SyntaxAnnotation s_newLocalDeclarationStatementAnnotation = new SyntaxAnnotation();
+        private static readonly SyntaxAnnotation s_unusedLocalDeclarationAnnotation = new SyntaxAnnotation();
+
+        public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(IDEDiagnosticIds.ExpressionValueIsUnusedDiagnosticId,
+                                                                                                    IDEDiagnosticIds.ValueAssignedIsUnusedDiagnosticId);
```

Comment:
not quite clear the distinction between thse two guys.  Also... is there a need to share the same FixProvider for multiple different analyzers?  in genreal that makes things much more confusing as the code then has to specialize for the different cases, instead of having the entire type be the specialization for the case.

Reply:
`ExpressionValueIsUnusedDiagnosticId` is specifically for values dropped by expression statements, which has a different default severity (hidden/refactoring) and is non-fading versus `ValueAssignedIsUnusedDiagnosticId`, which is for unused values assigned to locals/parameters, which has default severity Suggestion and is fading. Both of them deal with unused values, but need different diagnostic ID/option due to the former being so special and common in existing code.

---
In reply to: [230604071](https://github.com/dotnet/roslyn/pull/30777#discussion_r230604071) [](ancestors = 230604071)

Reply:
> is there a need to share the same FixProvider for multiple different analyzers

The underlying code fix and unused value preference is very similar between these two diagnostics. Both of them share large number of language specific and language agnostic code in this type, include some post-processing. I had initially started with separate code fixers for them, but that led to lot of code duplication.

---
In reply to: [230824394](https://github.com/dotnet/roslyn/pull/30777#discussion_r230824394) [](ancestors = 230824394,230604071)

Reply:
having a common base type, with ID-specific subclasses, can be useful.  But i don't have a good enough sense of everything going on here to say if that would actually help or hinder here.